### ISSUE TITLE
Support for specifying file mode and ownership

### ISSUE THREAD

**Issue by hashbrowncipher** [hashicorp/consul-template#358]
Currently, consul-template's atomicWrite function preserves mode of any file it overwrites.  If a file doesn't exist, it will be created with mode 0644.  There is no support for specifying or preserving file ownership.  In #16, the possibility of specifying the permissions of output files was discussed.

My use-case is consul-template retrieving secrets from Vault.  Services with different security levels will coexist on a multi-tenant node, with each service having its own Linux user.  I would like to map a Vault policy to each service, and the whole node will have the union of all such policies.  consul-template will have access to the "union token", and will be responsible for generating tokens with a subset of its policies and distributing them to each service user.  Such files should be owned by serviceuser:servicegroup, and have mode no greater than 0640.  Writing data into the file and later changing the file permissions (as would happen using a post-write command) leaves a window of opportunity for an attacker to use a tool such as inotify to open the file before the post-write command changes its mode.

Back when #16 was being discussed, consul-template did not have Vault integration, and was not specialized to handling secrets.  Now that it does, I think it might be worth revisiting some of that discussion.

I can envision three possible solutions that would fit the above use-case:

1. provide a global option to specify the mode for all files created by consul-template.  This _could_ just be the umask of the consul-template process.  consul-template would write files into a staging area according to the configured value, and post-write commands would use cp, chown, and mv to place them in their final destinations with the appropriate permissions.
2. same as 1, but _also_ preserve ownership of existing files.  This mostly eliminates the need for a staging area, meaning that post-write commands for each template need only do chown.
3. individually specify file ownership and mode in consul-template.hcl for each templated file, and set those attributes before any data is written into the file.

Is #16 worth revisiting? Do any of the above options make sense?

cc @EvanKrall, @bgirardeau

**Comment by sethvargo** [comment]
Hi @hashbrowncipher 

Thank you for opening an issue. I guess I don't see the reason you can't specify the file permissions as part of the command execution or just create the file with the correct permissions before starting CT?

I do suppose there is a security vulnerability in that we don't chmod the file until after the contents have been written, but we ran into issues where the file permissions on the target were too strict such that even CT couldn't write to them.

My counter-argument to your `inotify` command is that a skilled attacker could also just dump the memory of CT (or the Vault client) and inspect the secrets in plain-text. If someone has access to the system that is running CT, that's always going to be a vulnerability (it's also clearly called out in Vault's threat model docs).

Option 2 does not work - that was the original implementation and it does not work on Windows or network volumes; that's one of the big reasons we switched to an atomic write.

Option 3 seems reasonable to me, but it still poses the initial problem of "what if the permissions are too high such that even CT can't write to the file?"

I suppose the most viable option is to chmod the file before writing any data to it, but as I mentioned above, I don't think that's solving any security concerns. If you're able to dump the memory of the CT process (which is pretty trivial given access to the system), the secrets will be in plain-text.

Thoughts?

**Comment by hashbrowncipher** [comment]
Wow, what a quick response!

The threat model that I wish to defend against is an unprivileged (and possibly namespaced or jailed) Linux user gaining access to secrets that they are not entitled to.  Barring a vulnerability in Linux, such a user would not be able to look inside (by memory dump, ptrace, etc.) the consul-template process, but would be able to read new files as they are created by consul-template (possibly using inotify to know exactly when to open them).

I confess I'm a bit confused by the situations you mention where consul-template creates files with permissions so strict that it cannot write to them again.  Possibly you're considering a platform other than one that I'm familiar with? (you mentioned Windows, and I'm pretty Linux centric).  I've copied an example below where I've emulated a situation similar to one that I think consul-template would face.  I use Python's os.rename to rule out any "smarts" that mv might have.

```
❯❯❯ touch hi bye
❯❯❯ chmod 000 hi bye
❯❯❯ python2.7
Python 2.7.6 (default, Feb  4 2014, 16:58:51)
[GCC 4.4.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.rename('hi', 'bye')
>>>
```

So even with mode == 000, we would still be able to write a new version of the file.

Regarding creating the files ahead of time, I think it would successfully prevent an attacker from reading vulnerable secrets, but it introduces additional complexity that I have to maintain in order to assure that I have a secure system.  Namely, if I ever fail to create a file ahead of time, I disclose secrets from that file to every user on the machine.  I could see this happening if an administrator adds a file to consul-template.hcl but forgets to add it to the list of files that must be created beforehand.

**Comment by ryanuber** [comment]
I think that using a staging directory + the post command should be sufficient in this case. If the staging directory itself has restrictive permissions (0700 or so), any unprivileged user should not be able to reach in and read the file. This gives you the opportunity to chown/chmod and then mv the file to its final destination. Would that work for your case, @hashbrowncipher? Trying to support file ownership + permissions across both Linux and Windows might turn into a bigger pain than it seems.

**Comment by bgirardeau** [comment]
Either the staging directory or having existing files with the correct mode works, but a small change would make this easier and cleaner for users.

Since the temporary file is created with permissions 0600 (see line 55 [here](https://golang.org/src/io/ioutil/tempfile.go)), it isn't a concern that the chmod happens after writing the contents. The problem is that the only way to avoid the chmod temporarily setting permissions to 0644 is to have a file already there with the desired mode (which is error prone as @hashbrowncipher mentioned). 

Instead it would be nice to have the option to specify either a per-template mode, or a default mode for all templates instead of the hardcoded 0644 in atomicWrite. This gives users the option to keep sensitive information readable only by CT until explicitly asked through a chown/chmod post command, while not changing any behavior in the default case. Thoughts?

**Comment by hashbrowncipher** [comment]
I've been thinking about this some more.  Is it possible for each template definition to specify a create_command, just as it specifies a post-update command?  The contract of create_command would be to create a file for consul-template to write to, writing the name of the created file to stdout, very much in the vein of mktemp.  For this purpose, I would make my create_command a shell script which does something like:

```
#!/bin/bash
set -o errexit

file=$(mktemp "${1}.XXXXX")
chmod "${2}" "$file"
echo "$file"
```

**Comment by wwalker** [comment]
+1 for a create_command, or a pre_command

and passing the filename of the template to both of them would be nice, either as an argument, or an environment variable.
